java/ql/src/experimental/Security/CWE/CWE-470/UnsafeReflectionLib.qll

```diff
@@ -0,0 +1,94 @@
+import java
+import DataFlow
+import semmle.code.java.Reflection
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.DataFlow3
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * A call to a Java standard library method which constructs or returns a `Class<T>` from a `String`.
+ * e.g `Class.forName(...)` or `ClassLoader.loadClass(...)`
+ */
+class ReflectiveClassIdentifierMethodAccessCall extends MethodAccess {
+  ReflectiveClassIdentifierMethodAccessCall() {
+    this instanceof ReflectiveClassIdentifierMethodAccess
+  }
+}
+
+/**
+ * Unsafe reflection sink.
+ * e.g `Constructor.newInstance(...)` or `Method.invoke(...)` or `Class.newInstance()`.
+ */
+class UnsafeReflectionSink extends DataFlow::ExprNode {
+  UnsafeReflectionSink() {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Constructor<>", "newInstance")
+        or
+        ma.getMethod().hasQualifiedName("java.lang.reflect", "Method", "invoke")
+      ) and
+      ma.getQualifier() = this.asExpr() and
+      exists(ReflectionArgsConfig rac | rac.hasFlowToExpr(ma.getAnArgument()))
+    )
+  }
+}
+
+/** Taint-tracking configuration tracing flow from remote sources to specifying the initialization parameters to the constructor or method. */
+class ReflectionArgsConfig extends TaintTracking2::Configuration {
+  ReflectionArgsConfig() { this = "ReflectionArgsConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(NewInstance ni | ni.getAnArgument() = sink.asExpr())
+    or
+    exists(MethodAccess ma, ReflectionInvokeObjectConfig rioc |
+      ma.getMethod().hasQualifiedName("java.lang.reflect", "Method", "invoke") and
+      ma.getArgument(1) = sink.asExpr() and
+      rioc.hasFlow(_, DataFlow::exprNode(ma.getArgument(0)))
```

Comment:
You want both the method and the arg list to contain something user-controlled? Why?

Reply:
There needs to be configuration to ensure that the class object that calls the method through reflection is also controlled by the user. For example: the object in `method.invoke(object, parameter)`